coderd/prometheusmetrics/aggregator.go

```diff
@@ -0,0 +1,250 @@
+package prometheusmetrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/codersdk/agentsdk"
+)
+
+const (
+	// MetricHelpForAgent is a help string that replaces all agent metric help
+	// messages. This is because a registry cannot have conflicting
+	// help messages for the same metric in a "gather". If our coder agents are
+	// on different versions, this is a possible scenario.
+	metricHelpForAgent = "Metrics are forwarded from workspace agents connected to this instance of coderd."
+)
+
+const (
+	sizeCollectCh = 10
+	sizeUpdateCh  = 1024
+
+	defaultMetricsCleanupInterval = 2 * time.Minute
+)
+
+type MetricsAggregator struct {
+	queue []annotatedMetric
+
+	log                    slog.Logger
+	metricsCleanupInterval time.Duration
+
+	collectCh chan (chan []prometheus.Metric)
+	updateCh  chan updateRequest
+
+	updateHistogram  prometheus.Histogram
+	cleanupHistogram prometheus.Histogram
+}
+
+type updateRequest struct {
+	username      string
+	workspaceName string
+	agentName     string
+
+	metrics []agentsdk.AgentMetric
+
+	timestamp time.Time
+}
+
+type annotatedMetric struct {
+	agentsdk.AgentMetric
+
+	username      string
+	workspaceName string
+	agentName     string
+
+	expiryDate time.Time
+}
+
+var _ prometheus.Collector = new(MetricsAggregator)
+
+func NewMetricsAggregator(logger slog.Logger, registerer prometheus.Registerer, duration time.Duration) (*MetricsAggregator, error) {
+	metricsCleanupInterval := defaultMetricsCleanupInterval
+	if duration > 0 {
+		metricsCleanupInterval = duration
+	}
+
+	updateHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "coderd",
+		Subsystem: "prometheusmetrics",
+		Name:      "metrics_aggregator_execution_update_seconds",
+		Help:      "Histogram for duration of metrics aggregator update in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	err := registerer.Register(updateHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanupHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "coderd",
+		Subsystem: "prometheusmetrics",
+		Name:      "metrics_aggregator_execution_cleanup_seconds",
+		Help:      "Histogram for duration of metrics aggregator cleanup in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	err = registerer.Register(cleanupHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricsAggregator{
+		log:                    logger,
+		metricsCleanupInterval: metricsCleanupInterval,
+
+		collectCh: make(chan (chan []prometheus.Metric)),
```

Comment:
I think this will still need a buffer capacity, otherwise we might see the queue full error quite often.

Reply:
Good catch. I removed it accidentally.